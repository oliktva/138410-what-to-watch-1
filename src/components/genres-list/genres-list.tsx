import React, {FunctionComponent} from 'react';
import {Dispatch} from 'redux';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';

import {State, ActionCreator, Action} from 'src/reducer';
import {GenreProps, genrePropTypes, genresPropTypes} from 'src/types/genres';

interface StateProps {
  active: GenreProps;
}

interface DispatchProps {
  setFilterByGenre: (genre: GenreProps) => void;
}

interface OwnProps {
  genres: GenreProps[];
}

type Props = StateProps & DispatchProps & OwnProps;

const GenresList = (props: Props): JSX.Element => {
  const {genres, active, setFilterByGenre} = props;

  return (
    <ul className="catalog__genres-list">
      {genres.map((genre: GenreProps): JSX.Element => {
        const className = genre === active ?
          `catalog__genres-item catalog__genres-item--active` :
          `catalog__genres-item`;

        return (
          <li className={className} key={genre} onClick={(): void => setFilterByGenre(genre)}>
            <span className="catalog__genres-link">{genre}</span>
          </li>
        );
      })}
    </ul>
  );
};

const mapStateToProps = (state: State): StateProps => ({
  active: state.genre
});

const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => ({
  setFilterByGenre: (genre: GenreProps): Action => dispatch(ActionCreator.setFilterByGenre(genre))
});

GenresList.propTypes = {
  genres: genresPropTypes.isRequired,
  active: genrePropTypes.isRequired,
  setFilterByGenre: PropTypes.func.isRequired
};

export {GenresList};

const ConnectedComponent: any =
 connect<StateProps, DispatchProps, OwnProps, State>(mapStateToProps, mapDispatchToProps)(GenresList);

export default ConnectedComponent as FunctionComponent<OwnProps>;