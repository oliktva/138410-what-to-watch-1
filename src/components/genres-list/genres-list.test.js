import React from 'react';
import renderer from 'react-test-renderer';

import {GenresList} from './genres-list';

const genres = [`All genres`, `Drama`, `Adventure`, `Thriller`];

it(`renders correctly`, () => {
  const tree = renderer
    .create(<GenresList
      genres={genres}
      active="All genres"
      setFilterByGenre={jest.fn()}
    />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
