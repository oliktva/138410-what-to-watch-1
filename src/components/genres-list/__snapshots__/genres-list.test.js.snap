// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<ul
  className="catalog__genres-list"
>
  <li
    className="catalog__genres-item catalog__genres-item--active"
    key="All genres"
    onClick={[Function]}
  >
    <span
      className="catalog__genres-link"
    >
      All genres
    </span>
  </li>
  <li
    className="catalog__genres-item"
    key="Drama"
    onClick={[Function]}
  >
    <span
      className="catalog__genres-link"
    >
      Drama
    </span>
  </li>
  <li
    className="catalog__genres-item"
    key="Adventure"
    onClick={[Function]}
  >
    <span
      className="catalog__genres-link"
    >
      Adventure
    </span>
  </li>
  <li
    className="catalog__genres-item"
    key="Thriller"
    onClick={[Function]}
  >
    <span
      className="catalog__genres-link"
    >
      Thriller
    </span>
  </li>
</ul>
`;
